File: app/src/main/java/com/rafaelfiume/prosciutto/adviser/integration/ProductAdviserQuery.java
package com.rafaelfiume.prosciutto.adviser.integration;

import com.rafaelfiume.prosciutto.adviser.domain.Product;
import com.rafaelfiume.prosciutto.adviser.ProductAdviserParser;

import org.apache.commons.io.IOUtils;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.List;

import static com.rafaelfiume.prosciutto.adviser.integration.Get.get;

public enum ProductAdviserQuery implements ProductAdviserServiceClient {

    MAGIC {
        @Override
        public List<Product> suggestedProducts() throws Exception {
            return retrieveProductsSuggestionsFor("Magic");
        }
    },

    HEALTHY {
        @Override
        public List<Product> suggestedProducts() throws Exception {
            return retrieveProductsSuggestionsFor("Healthy");
        }
    },

    GOURMET {
        @Override
        public List<Product> suggestedProducts() throws Exception {
            return retrieveProductsSuggestionsFor("Gourmet");
        }
    },

    // Missing integration tests for the other three methods above
    EXPERT {
        @Override
        public List<Product> suggestedProducts() throws Exception {
            return this.retrieveProductsSuggestionsFor("Expert");
        }
    };

    protected List<Product> retrieveProductsSuggestionsFor(String profile) throws Exception {
        return parser.parse(get(ADVISER_WEB_SERVICE + profile));
    }

    private static final String ADVISER_WEB_SERVICE = ServerName.name() + "/advise/for/";

    private final ProductAdviserParser parser = new ProductAdviserParser();

}

File: app/src/main/java/com/rafaelfiume/prosciutto/adviser/integration/ProductAdviserQuery.kt
package com.rafaelfiume.prosciutto.adviser.integration

import com.rafaelfiume.prosciutto.adviser.ProductAdviserParser
import com.rafaelfiume.prosciutto.adviser.domain.Product
import com.rafaelfiume.prosciutto.adviser.integration.Get.get

enum class ProductAdviserQuery : ProductAdviserServiceClient {

    MAGIC {
        @Throws(Exception::class)
        override fun suggestedProducts(): List<Product> {
            return retrieveProductsSuggestionsFor("Magic")
        }
    },

    HEALTHY {
        @Throws(Exception::class)
        override fun suggestedProducts(): List<Product> {
            return retrieveProductsSuggestionsFor("Healthy")
        }
    },

    GOURMET {
        @Throws(Exception::class)
        override fun suggestedProducts(): List<Product> {
            return retrieveProductsSuggestionsFor("Gourmet")
        }
    },

    // Missing integration tests for the other three methods above
    EXPERT {
        @Throws(Exception::class)
        override fun suggestedProducts(): List<Product> {
            return this.retrieveProductsSuggestionsFor("Expert")
        }
    };

    @Throws(Exception::class)
    protected fun retrieveProductsSuggestionsFor(profile: String): List<Product> {
        return parser.parse(get(ADVISER_WEB_SERVICE + profile))
    }

    private val parser = ProductAdviserParser()

    companion object {

        private val ADVISER_WEB_SERVICE = ServerName.name() + "/advise/for/"
    }

}
